lib/libvfio-user.c

```diff
@@ -1052,6 +1052,45 @@ handle_dirty_pages(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
     return ret;
 }
 
+static int
+handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+{
+    struct vfio_user_device_feature *req = msg->in.iov.iov_base;
+
+    if (vfu_ctx->migration == NULL) {
+        return -EINVAL;
+    }
+
+    if (!migration_feature_supported(req->flags)) {
+        // FIXME what error code to return? we really want "not supported"
+        //   instead of "not permitted"?
+        return -EINVAL;
+    }
+
+    ssize_t ret;
+
+    if (req->flags & VFIO_DEVICE_FEATURE_PROBE) {
+        msg->out.iov.iov_base = msg->in.iov.iov_base;
+        msg->out.iov.iov_len = msg->in.iov.iov_len;
+
+        ret = 0;
+    } else if (req->flags & VFIO_DEVICE_FEATURE_GET) {
+        msg->out.iov.iov_base = calloc(8, 1);
```

Comment:
1, 8 is more typical. You don't check for allocation failure.
If you set iov_len first, you can then use that value and not erpeat yourself.
Finally, I'd use sizeof(vfio_user_device_feature_migration), with a comment that currently all supported outgoing data is the same size as this struct (8 bytes).